// @ai-sdk/alibaba/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  VERSION: () => VERSION,
  alibaba: () => alibaba,
  createAlibaba: () => createAlibaba
});
module.exports = __toCommonJS(index_exports);

// src/alibaba-provider.ts
var import_provider4 = require("@ai-sdk/provider");
var import_provider_utils5 = require("@ai-sdk/provider-utils");

// src/alibaba-chat-language-model.ts
var import_internal2 = require("@ai-sdk/openai-compatible/internal");
var import_provider2 = require("@ai-sdk/provider");
var import_provider_utils3 = require("@ai-sdk/provider-utils");
var import_v43 = require("zod/v4");

// src/alibaba-chat-options.ts
var import_v4 = require("zod/v4");
var alibabaLanguageModelOptions = import_v4.z.object({
  /**
   * Enable thinking/reasoning mode for supported models.
   * When enabled, the model generates reasoning content before the response.
   *
   * @default false
   */
  enableThinking: import_v4.z.boolean().optional(),
  /**
   * Maximum number of reasoning tokens to generate.
   */
  thinkingBudget: import_v4.z.number().positive().optional(),
  /**
   * Whether to enable parallel function calling during tool use.
   *
   * @default true
   */
  parallelToolCalls: import_v4.z.boolean().optional()
});

// src/alibaba-error.ts
var import_provider_utils = require("@ai-sdk/provider-utils");
var import_v42 = require("zod/v4");
var alibabaErrorDataSchema = import_v42.z.object({
  error: import_v42.z.object({
    message: import_v42.z.string(),
    code: import_v42.z.string().nullish(),
    type: import_v42.z.string().nullish()
  })
});
var alibabaFailedResponseHandler = (0, import_provider_utils.createJsonErrorResponseHandler)({
  errorSchema: alibabaErrorDataSchema,
  errorToMessage: (data) => data.error.message
});

// src/convert-alibaba-usage.ts
var import_internal = require("@ai-sdk/openai-compatible/internal");
function convertAlibabaUsage(usage) {
  var _a, _b, _c, _d;
  const baseUsage = (0, import_internal.convertOpenAICompatibleChatUsage)(usage);
  const cacheWriteTokens = (_b = (_a = usage == null ? void 0 : usage.prompt_tokens_details) == null ? void 0 : _a.cache_creation_input_tokens) != null ? _b : 0;
  const noCacheTokens = ((_c = baseUsage.inputTokens.total) != null ? _c : 0) - ((_d = baseUsage.inputTokens.cacheRead) != null ? _d : 0) - cacheWriteTokens;
  return {
    ...baseUsage,
    inputTokens: {
      ...baseUsage.inputTokens,
      cacheWrite: cacheWriteTokens,
      noCache: noCacheTokens
    }
  };
}

// src/convert-to-alibaba-chat-messages.ts
var import_provider = require("@ai-sdk/provider");
var import_provider_utils2 = require("@ai-sdk/provider-utils");
function formatImageUrl({
  data,
  mediaType
}) {
  return data instanceof URL ? data.toString() : `data:${mediaType};base64,${(0, import_provider_utils2.convertToBase64)(data)}`;
}
function convertToAlibabaChatMessages({
  prompt,
  cacheControlValidator
}) {
  var _a, _b;
  const messages = [];
  for (const { role, content, ...message } of prompt) {
    const messageCacheControl = cacheControlValidator == null ? void 0 : cacheControlValidator.getCacheControl(
      message.providerOptions
    );
    switch (role) {
      case "system": {
        if (messageCacheControl) {
          messages.push({
            role: "system",
            content: [
              {
                type: "text",
                text: content,
                cache_control: messageCacheControl
              }
            ]
          });
        } else {
          messages.push({ role: "system", content });
        }
        break;
      }
      case "user": {
        messages.push({
          role: "user",
          content: content.map((part, index) => {
            var _a2;
            const isLastPart = index === content.length - 1;
            const partCacheControl = (_a2 = cacheControlValidator == null ? void 0 : cacheControlValidator.getCacheControl(part.providerOptions)) != null ? _a2 : isLastPart ? messageCacheControl : void 0;
            switch (part.type) {
              case "text": {
                return {
                  type: "text",
                  text: part.text,
                  ...partCacheControl ? { cache_control: partCacheControl } : {}
                };
              }
              case "file": {
                if (part.mediaType.startsWith("image/")) {
                  const mediaType = part.mediaType === "image/*" ? "image/jpeg" : part.mediaType;
                  return {
                    type: "image_url",
                    image_url: {
                      url: formatImageUrl({ data: part.data, mediaType })
                    },
                    ...partCacheControl ? { cache_control: partCacheControl } : {}
                  };
                } else {
                  throw new import_provider.UnsupportedFunctionalityError({
                    functionality: "Only image file parts are supported"
                  });
                }
              }
            }
          })
        });
        break;
      }
      case "assistant": {
        let text = "";
        const toolCalls = [];
        for (const part of content) {
          switch (part.type) {
            case "text": {
              text += part.text;
              break;
            }
            case "tool-call": {
              toolCalls.push({
                id: part.toolCallId,
                type: "function",
                function: {
                  name: part.toolName,
                  arguments: JSON.stringify(part.input)
                }
              });
              break;
            }
            case "reasoning": {
              text += part.text;
              break;
            }
          }
        }
        messages.push({
          role: "assistant",
          content: messageCacheControl ? [{ type: "text", text, cache_control: messageCacheControl }] : text || null,
          tool_calls: toolCalls.length > 0 ? toolCalls : void 0
        });
        break;
      }
      case "tool": {
        const toolResponses = content.filter(
          (r) => r.type !== "tool-approval-response"
        );
        for (let i = 0; i < toolResponses.length; i++) {
          const toolResponse = toolResponses[i];
          const output = toolResponse.output;
          const isLastPart = i === toolResponses.length - 1;
          const partCacheControl = (_a = cacheControlValidator == null ? void 0 : cacheControlValidator.getCacheControl(
            toolResponse.providerOptions
          )) != null ? _a : isLastPart ? messageCacheControl : void 0;
          let contentValue;
          switch (output.type) {
            case "text":
            case "error-text":
              contentValue = output.value;
              break;
            case "execution-denied":
              contentValue = (_b = output.reason) != null ? _b : "Tool execution denied.";
              break;
            case "content":
            case "json":
            case "error-json":
              contentValue = JSON.stringify(output.value);
              break;
          }
          messages.push({
            role: "tool",
            tool_call_id: toolResponse.toolCallId,
            content: partCacheControl ? [
              {
                type: "text",
                text: contentValue,
                cache_control: partCacheControl
              }
            ] : contentValue
          });
        }
        break;
      }
      default: {
        const _exhaustiveCheck = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }
  return messages;
}

// src/get-cache-control.ts
var MAX_CACHE_BREAKPOINTS = 4;
function getCacheControl(providerMetadata) {
  var _a;
  const alibaba2 = providerMetadata == null ? void 0 : providerMetadata.alibaba;
  const cacheControlValue = (_a = alibaba2 == null ? void 0 : alibaba2.cacheControl) != null ? _a : alibaba2 == null ? void 0 : alibaba2.cache_control;
  return cacheControlValue;
}
var CacheControlValidator = class {
  constructor() {
    this.breakpointCount = 0;
    this.warnings = [];
  }
  getCacheControl(providerMetadata) {
    const cacheControlValue = getCacheControl(providerMetadata);
    if (!cacheControlValue) {
      return void 0;
    }
    this.breakpointCount++;
    if (this.breakpointCount > MAX_CACHE_BREAKPOINTS) {
      this.warnings.push({
        type: "other",
        message: `Max breakpoint limit exceeded. Only the last ${MAX_CACHE_BREAKPOINTS} cache markers will take effect.`
      });
    }
    return cacheControlValue;
  }
  getWarnings() {
    return this.warnings;
  }
};

// src/alibaba-chat-language-model.ts
var AlibabaLanguageModel = class {
  constructor(modelId, config) {
    this.specificationVersion = "v3";
    this.supportedUrls = {
      "image/*": [/^https?:\/\/.*$/]
    };
    this.modelId = modelId;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  /**
   * Builds request arguments for Alibaba API call.
   * Converts AI SDK options to Alibaba API format.
   */
  async getArgs({
    prompt,
    maxOutputTokens,
    temperature,
    topP,
    topK,
    frequencyPenalty,
    presencePenalty,
    stopSequences,
    responseFormat,
    seed,
    providerOptions,
    tools,
    toolChoice
  }) {
    var _a;
    const warnings = [];
    const cacheControlValidator = new CacheControlValidator();
    const alibabaOptions = await (0, import_provider_utils3.parseProviderOptions)({
      provider: "alibaba",
      providerOptions,
      schema: alibabaLanguageModelOptions
    });
    if (frequencyPenalty != null) {
      warnings.push({ type: "unsupported", feature: "frequencyPenalty" });
    }
    const baseArgs = {
      model: this.modelId,
      max_tokens: maxOutputTokens,
      temperature,
      top_p: topP,
      top_k: topK,
      presence_penalty: presencePenalty,
      stop: stopSequences,
      seed,
      response_format: (responseFormat == null ? void 0 : responseFormat.type) === "json" ? responseFormat.schema != null ? {
        type: "json_schema",
        json_schema: {
          schema: responseFormat.schema,
          name: (_a = responseFormat.name) != null ? _a : "response",
          description: responseFormat.description
        }
      } : { type: "json_object" } : void 0,
      // Alibaba-specific options
      ...(alibabaOptions == null ? void 0 : alibabaOptions.enableThinking) != null ? { enable_thinking: alibabaOptions.enableThinking } : {},
      ...(alibabaOptions == null ? void 0 : alibabaOptions.thinkingBudget) != null ? { thinking_budget: alibabaOptions.thinkingBudget } : {},
      // Convert messages with cache control support
      messages: convertToAlibabaChatMessages({
        prompt,
        cacheControlValidator
      })
    };
    const {
      tools: alibabaTools,
      toolChoice: alibabaToolChoice,
      toolWarnings
    } = (0, import_internal2.prepareTools)({ tools, toolChoice });
    warnings.push(...cacheControlValidator.getWarnings());
    return {
      args: {
        ...baseArgs,
        tools: alibabaTools,
        tool_choice: alibabaToolChoice,
        ...alibabaTools != null && (alibabaOptions == null ? void 0 : alibabaOptions.parallelToolCalls) !== void 0 ? { parallel_tool_calls: alibabaOptions.parallelToolCalls } : {}
      },
      warnings: [...warnings, ...toolWarnings]
    };
  }
  async doGenerate(options) {
    var _a;
    const { args, warnings } = await this.getArgs(options);
    const {
      responseHeaders,
      value: response,
      rawValue: rawResponse
    } = await (0, import_provider_utils3.postJsonToApi)({
      url: `${this.config.baseURL}/chat/completions`,
      headers: (0, import_provider_utils3.combineHeaders)(this.config.headers(), options.headers),
      body: args,
      failedResponseHandler: alibabaFailedResponseHandler,
      successfulResponseHandler: (0, import_provider_utils3.createJsonResponseHandler)(
        alibabaChatResponseSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const choice = response.choices[0];
    const content = [];
    const text = choice.message.content;
    if (text != null && text.length > 0) {
      content.push({ type: "text", text });
    }
    const reasoning = choice.message.reasoning_content;
    if (reasoning != null && reasoning.length > 0) {
      content.push({
        type: "reasoning",
        text: reasoning
      });
    }
    if (choice.message.tool_calls != null) {
      for (const toolCall of choice.message.tool_calls) {
        content.push({
          type: "tool-call",
          toolCallId: toolCall.id,
          toolName: toolCall.function.name,
          input: toolCall.function.arguments
        });
      }
    }
    return {
      content,
      finishReason: {
        unified: (0, import_internal2.mapOpenAICompatibleFinishReason)(choice.finish_reason),
        raw: (_a = choice.finish_reason) != null ? _a : void 0
      },
      usage: convertAlibabaUsage(response.usage),
      request: { body: JSON.stringify(args) },
      response: {
        ...(0, import_internal2.getResponseMetadata)(response),
        headers: responseHeaders,
        body: rawResponse
      },
      warnings
    };
  }
  async doStream(options) {
    const { args, warnings } = await this.getArgs(options);
    const body = {
      ...args,
      stream: true,
      stream_options: this.config.includeUsage ? { include_usage: true } : void 0
    };
    const { responseHeaders, value: response } = await (0, import_provider_utils3.postJsonToApi)({
      url: `${this.config.baseURL}/chat/completions`,
      headers: (0, import_provider_utils3.combineHeaders)(this.config.headers(), options.headers),
      body,
      failedResponseHandler: alibabaFailedResponseHandler,
      successfulResponseHandler: (0, import_provider_utils3.createEventSourceResponseHandler)(
        alibabaChatChunkSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    let finishReason = {
      unified: "other",
      raw: void 0
    };
    let usage = void 0;
    let isFirstChunk = true;
    let activeText = false;
    let activeReasoningId = null;
    const toolCalls = [];
    return {
      stream: response.pipeThrough(
        new TransformStream({
          start(controller) {
            controller.enqueue({ type: "stream-start", warnings });
          },
          transform(chunk, controller) {
            var _a, _b, _c, _d;
            if (options.includeRawChunks) {
              controller.enqueue({ type: "raw", rawValue: chunk.rawValue });
            }
            if (!chunk.success) {
              controller.enqueue({ type: "error", error: chunk.error });
              return;
            }
            const value = chunk.value;
            if (isFirstChunk) {
              isFirstChunk = false;
              controller.enqueue({
                type: "response-metadata",
                ...(0, import_internal2.getResponseMetadata)(value)
              });
            }
            if (value.usage != null) {
              usage = value.usage;
            }
            if (value.choices.length === 0) {
              return;
            }
            const choice = value.choices[0];
            const delta = choice.delta;
            if (delta.reasoning_content != null && delta.reasoning_content.length > 0) {
              if (activeReasoningId == null) {
                if (activeText) {
                  controller.enqueue({ type: "text-end", id: "0" });
                  activeText = false;
                }
                activeReasoningId = (0, import_provider_utils3.generateId)();
                controller.enqueue({
                  type: "reasoning-start",
                  id: activeReasoningId
                });
              }
              controller.enqueue({
                type: "reasoning-delta",
                id: activeReasoningId,
                delta: delta.reasoning_content
              });
            }
            if (delta.content != null && delta.content.length > 0) {
              if (activeReasoningId != null) {
                controller.enqueue({
                  type: "reasoning-end",
                  id: activeReasoningId
                });
                activeReasoningId = null;
              }
              if (!activeText) {
                controller.enqueue({ type: "text-start", id: "0" });
                activeText = true;
              }
              controller.enqueue({
                type: "text-delta",
                id: "0",
                delta: delta.content
              });
            }
            if (delta.tool_calls != null) {
              if (activeReasoningId != null) {
                controller.enqueue({
                  type: "reasoning-end",
                  id: activeReasoningId
                });
                activeReasoningId = null;
              }
              if (activeText) {
                controller.enqueue({ type: "text-end", id: "0" });
                activeText = false;
              }
              for (const toolCallDelta of delta.tool_calls) {
                const index = (_a = toolCallDelta.index) != null ? _a : toolCalls.length;
                if (toolCalls[index] == null) {
                  if (toolCallDelta.id == null) {
                    throw new import_provider2.InvalidResponseDataError({
                      data: toolCallDelta,
                      message: `Expected 'id' to be a string.`
                    });
                  }
                  if (((_b = toolCallDelta.function) == null ? void 0 : _b.name) == null) {
                    throw new import_provider2.InvalidResponseDataError({
                      data: toolCallDelta,
                      message: `Expected 'function.name' to be a string.`
                    });
                  }
                  controller.enqueue({
                    type: "tool-input-start",
                    id: toolCallDelta.id,
                    toolName: toolCallDelta.function.name
                  });
                  toolCalls[index] = {
                    id: toolCallDelta.id,
                    type: "function",
                    function: {
                      name: toolCallDelta.function.name,
                      arguments: (_c = toolCallDelta.function.arguments) != null ? _c : ""
                    },
                    hasFinished: false
                  };
                  const toolCall2 = toolCalls[index];
                  if (toolCall2.function.arguments.length > 0) {
                    controller.enqueue({
                      type: "tool-input-delta",
                      id: toolCall2.id,
                      delta: toolCall2.function.arguments
                    });
                  }
                  if ((0, import_provider_utils3.isParsableJson)(toolCall2.function.arguments)) {
                    controller.enqueue({
                      type: "tool-input-end",
                      id: toolCall2.id
                    });
                    controller.enqueue({
                      type: "tool-call",
                      toolCallId: toolCall2.id,
                      toolName: toolCall2.function.name,
                      input: toolCall2.function.arguments
                    });
                    toolCall2.hasFinished = true;
                  }
                  continue;
                }
                const toolCall = toolCalls[index];
                if (toolCall.hasFinished) {
                  continue;
                }
                if (((_d = toolCallDelta.function) == null ? void 0 : _d.arguments) != null) {
                  toolCall.function.arguments += toolCallDelta.function.arguments;
                  controller.enqueue({
                    type: "tool-input-delta",
                    id: toolCall.id,
                    delta: toolCallDelta.function.arguments
                  });
                }
                if ((0, import_provider_utils3.isParsableJson)(toolCall.function.arguments)) {
                  controller.enqueue({
                    type: "tool-input-end",
                    id: toolCall.id
                  });
                  controller.enqueue({
                    type: "tool-call",
                    toolCallId: toolCall.id,
                    toolName: toolCall.function.name,
                    input: toolCall.function.arguments
                  });
                  toolCall.hasFinished = true;
                }
              }
            }
            if (choice.finish_reason != null) {
              finishReason = {
                unified: (0, import_internal2.mapOpenAICompatibleFinishReason)(choice.finish_reason),
                raw: choice.finish_reason
              };
            }
          },
          flush(controller) {
            if (activeReasoningId != null) {
              controller.enqueue({
                type: "reasoning-end",
                id: activeReasoningId
              });
            }
            if (activeText) {
              controller.enqueue({ type: "text-end", id: "0" });
            }
            controller.enqueue({
              type: "finish",
              finishReason,
              usage: convertAlibabaUsage(usage)
            });
          }
        })
      ),
      request: { body: JSON.stringify(body) },
      response: { headers: responseHeaders }
    };
  }
};
var alibabaUsageSchema = import_v43.z.object({
  prompt_tokens: import_v43.z.number(),
  completion_tokens: import_v43.z.number(),
  total_tokens: import_v43.z.number(),
  prompt_tokens_details: import_v43.z.object({
    cached_tokens: import_v43.z.number().nullish(),
    cache_creation_input_tokens: import_v43.z.number().nullish()
  }).nullish(),
  completion_tokens_details: import_v43.z.object({
    reasoning_tokens: import_v43.z.number().nullish()
  }).nullish()
});
var alibabaChatResponseSchema = import_v43.z.object({
  id: import_v43.z.string().nullish(),
  created: import_v43.z.number().nullish(),
  model: import_v43.z.string().nullish(),
  choices: import_v43.z.array(
    import_v43.z.object({
      message: import_v43.z.object({
        role: import_v43.z.literal("assistant").nullish(),
        content: import_v43.z.string().nullish(),
        reasoning_content: import_v43.z.string().nullish(),
        // Alibaba thinking mode
        tool_calls: import_v43.z.array(
          import_v43.z.object({
            id: import_v43.z.string(),
            type: import_v43.z.literal("function"),
            function: import_v43.z.object({
              name: import_v43.z.string(),
              arguments: import_v43.z.string()
            })
          })
        ).nullish()
      }),
      finish_reason: import_v43.z.string().nullish(),
      index: import_v43.z.number()
    })
  ),
  usage: alibabaUsageSchema.nullish()
});
var alibabaChatChunkSchema = import_v43.z.object({
  id: import_v43.z.string().nullish(),
  created: import_v43.z.number().nullish(),
  model: import_v43.z.string().nullish(),
  choices: import_v43.z.array(
    import_v43.z.object({
      delta: import_v43.z.object({
        role: import_v43.z.enum(["assistant"]).nullish(),
        content: import_v43.z.string().nullish(),
        reasoning_content: import_v43.z.string().nullish(),
        // Alibaba thinking mode delta
        tool_calls: import_v43.z.array(
          import_v43.z.object({
            index: import_v43.z.number().nullish(),
            // Index for accumulating tool calls
            id: import_v43.z.string().nullish(),
            type: import_v43.z.literal("function").nullish(),
            function: import_v43.z.object({
              name: import_v43.z.string().nullish(),
              arguments: import_v43.z.string().nullish()
            }).nullish()
          })
        ).nullish()
      }),
      finish_reason: import_v43.z.string().nullish(),
      index: import_v43.z.number()
    })
  ),
  usage: alibabaUsageSchema.nullish()
  // Usage only appears in final chunk
});

// src/alibaba-video-model.ts
var import_provider3 = require("@ai-sdk/provider");
var import_provider_utils4 = require("@ai-sdk/provider-utils");
var import_v44 = require("zod/v4");
var alibabaVideoModelOptionsSchema = (0, import_provider_utils4.lazySchema)(
  () => (0, import_provider_utils4.zodSchema)(
    import_v44.z.object({
      negativePrompt: import_v44.z.string().nullish(),
      audioUrl: import_v44.z.string().nullish(),
      promptExtend: import_v44.z.boolean().nullish(),
      shotType: import_v44.z.enum(["single", "multi"]).nullish(),
      watermark: import_v44.z.boolean().nullish(),
      audio: import_v44.z.boolean().nullish(),
      referenceUrls: import_v44.z.array(import_v44.z.string()).nullish(),
      pollIntervalMs: import_v44.z.number().positive().nullish(),
      pollTimeoutMs: import_v44.z.number().positive().nullish()
    }).passthrough()
  )
);
var alibabaVideoErrorSchema = import_v44.z.object({
  code: import_v44.z.string().nullish(),
  message: import_v44.z.string(),
  request_id: import_v44.z.string().nullish()
});
var alibabaVideoFailedResponseHandler = (0, import_provider_utils4.createJsonErrorResponseHandler)({
  errorSchema: alibabaVideoErrorSchema,
  errorToMessage: (data) => data.message
});
var alibabaVideoCreateTaskSchema = import_v44.z.object({
  output: import_v44.z.object({
    task_status: import_v44.z.string(),
    task_id: import_v44.z.string()
  }).nullish(),
  request_id: import_v44.z.string().nullish()
});
var alibabaVideoTaskStatusSchema = import_v44.z.object({
  output: import_v44.z.object({
    task_id: import_v44.z.string(),
    task_status: import_v44.z.string(),
    video_url: import_v44.z.string().nullish(),
    submit_time: import_v44.z.string().nullish(),
    scheduled_time: import_v44.z.string().nullish(),
    end_time: import_v44.z.string().nullish(),
    orig_prompt: import_v44.z.string().nullish(),
    actual_prompt: import_v44.z.string().nullish(),
    code: import_v44.z.string().nullish(),
    message: import_v44.z.string().nullish()
  }).nullish(),
  usage: import_v44.z.object({
    duration: import_v44.z.number().nullish(),
    output_video_duration: import_v44.z.number().nullish(),
    SR: import_v44.z.number().nullish(),
    size: import_v44.z.string().nullish()
  }).nullish(),
  request_id: import_v44.z.string().nullish()
});
function detectMode(modelId) {
  if (modelId.includes("-i2v")) return "i2v";
  if (modelId.includes("-r2v")) return "r2v";
  return "t2v";
}
var AlibabaVideoModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v3";
    this.maxVideosPerCall = 1;
  }
  get provider() {
    return this.config.provider;
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k;
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const warnings = [];
    const mode = detectMode(this.modelId);
    const alibabaOptions = await (0, import_provider_utils4.parseProviderOptions)({
      provider: "alibaba",
      providerOptions: options.providerOptions,
      schema: alibabaVideoModelOptionsSchema
    });
    const input = {};
    if (options.prompt != null) {
      input.prompt = options.prompt;
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.negativePrompt) != null) {
      input.negative_prompt = alibabaOptions.negativePrompt;
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.audioUrl) != null) {
      input.audio_url = alibabaOptions.audioUrl;
    }
    if (mode === "i2v" && options.image != null) {
      if (options.image.type === "url") {
        input.img_url = options.image.url;
      } else {
        const base64Data = typeof options.image.data === "string" ? options.image.data : (0, import_provider_utils4.convertUint8ArrayToBase64)(options.image.data);
        input.img_url = base64Data;
      }
    }
    if (mode === "r2v" && (alibabaOptions == null ? void 0 : alibabaOptions.referenceUrls) != null) {
      input.reference_urls = alibabaOptions.referenceUrls;
    }
    const parameters = {};
    if (options.duration != null) {
      parameters.duration = options.duration;
    }
    if (options.seed != null) {
      parameters.seed = options.seed;
    }
    if (options.resolution != null) {
      if (mode === "i2v") {
        const resolutionMap = {
          "1280x720": "720P",
          "720x1280": "720P",
          "960x960": "720P",
          "1088x832": "720P",
          "832x1088": "720P",
          "1920x1080": "1080P",
          "1080x1920": "1080P",
          "1440x1440": "1080P",
          "1632x1248": "1080P",
          "1248x1632": "1080P",
          "832x480": "480P",
          "480x832": "480P",
          "624x624": "480P"
        };
        parameters.resolution = resolutionMap[options.resolution] || options.resolution;
      } else {
        parameters.size = options.resolution.replace("x", "*");
      }
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.promptExtend) != null) {
      parameters.prompt_extend = alibabaOptions.promptExtend;
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.shotType) != null) {
      parameters.shot_type = alibabaOptions.shotType;
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.watermark) != null) {
      parameters.watermark = alibabaOptions.watermark;
    }
    if ((alibabaOptions == null ? void 0 : alibabaOptions.audio) != null) {
      parameters.audio = alibabaOptions.audio;
    }
    if (options.aspectRatio) {
      warnings.push({
        type: "unsupported",
        feature: "aspectRatio",
        details: "Alibaba video models use explicit size/resolution dimensions. Use the resolution option or providerOptions.alibaba for size control."
      });
    }
    if (options.fps) {
      warnings.push({
        type: "unsupported",
        feature: "fps",
        details: "Alibaba video models do not support custom FPS."
      });
    }
    if (options.n != null && options.n > 1) {
      warnings.push({
        type: "unsupported",
        feature: "n",
        details: "Alibaba video models only support generating 1 video per call."
      });
    }
    const { value: createResponse } = await (0, import_provider_utils4.postJsonToApi)({
      url: `${this.config.baseURL}/api/v1/services/aigc/video-generation/video-synthesis`,
      headers: (0, import_provider_utils4.combineHeaders)(
        await (0, import_provider_utils4.resolve)(this.config.headers),
        options.headers,
        {
          "X-DashScope-Async": "enable"
        }
      ),
      body: {
        model: this.modelId,
        input,
        parameters
      },
      successfulResponseHandler: (0, import_provider_utils4.createJsonResponseHandler)(
        alibabaVideoCreateTaskSchema
      ),
      failedResponseHandler: alibabaVideoFailedResponseHandler,
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const taskId = (_d = createResponse.output) == null ? void 0 : _d.task_id;
    if (!taskId) {
      throw new import_provider3.AISDKError({
        name: "ALIBABA_VIDEO_GENERATION_ERROR",
        message: `No task_id returned from Alibaba API. Response: ${JSON.stringify(createResponse)}`
      });
    }
    const pollIntervalMs = (_e = alibabaOptions == null ? void 0 : alibabaOptions.pollIntervalMs) != null ? _e : 5e3;
    const pollTimeoutMs = (_f = alibabaOptions == null ? void 0 : alibabaOptions.pollTimeoutMs) != null ? _f : 6e5;
    const startTime = Date.now();
    let finalResponse;
    let responseHeaders;
    while (true) {
      await (0, import_provider_utils4.delay)(pollIntervalMs, { abortSignal: options.abortSignal });
      if (Date.now() - startTime > pollTimeoutMs) {
        throw new import_provider3.AISDKError({
          name: "ALIBABA_VIDEO_GENERATION_TIMEOUT",
          message: `Video generation timed out after ${pollTimeoutMs}ms`
        });
      }
      const { value: statusResponse, responseHeaders: pollHeaders } = await (0, import_provider_utils4.getFromApi)({
        url: `${this.config.baseURL}/api/v1/tasks/${taskId}`,
        headers: (0, import_provider_utils4.combineHeaders)(
          await (0, import_provider_utils4.resolve)(this.config.headers),
          options.headers
        ),
        successfulResponseHandler: (0, import_provider_utils4.createJsonResponseHandler)(
          alibabaVideoTaskStatusSchema
        ),
        failedResponseHandler: alibabaVideoFailedResponseHandler,
        abortSignal: options.abortSignal,
        fetch: this.config.fetch
      });
      responseHeaders = pollHeaders;
      const taskStatus = (_g = statusResponse.output) == null ? void 0 : _g.task_status;
      if (taskStatus === "SUCCEEDED") {
        finalResponse = statusResponse;
        break;
      }
      if (taskStatus === "FAILED" || taskStatus === "CANCELED") {
        throw new import_provider3.AISDKError({
          name: "ALIBABA_VIDEO_GENERATION_FAILED",
          message: `Video generation ${taskStatus.toLowerCase()}. Task ID: ${taskId}. ${(_i = (_h = statusResponse.output) == null ? void 0 : _h.message) != null ? _i : ""}`
        });
      }
    }
    const videoUrl = (_j = finalResponse == null ? void 0 : finalResponse.output) == null ? void 0 : _j.video_url;
    if (!videoUrl) {
      throw new import_provider3.AISDKError({
        name: "ALIBABA_VIDEO_GENERATION_ERROR",
        message: `No video URL in response. Task ID: ${taskId}`
      });
    }
    return {
      videos: [
        {
          type: "url",
          url: videoUrl,
          mediaType: "video/mp4"
        }
      ],
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      },
      providerMetadata: {
        alibaba: {
          taskId,
          videoUrl,
          ...((_k = finalResponse == null ? void 0 : finalResponse.output) == null ? void 0 : _k.actual_prompt) ? { actualPrompt: finalResponse.output.actual_prompt } : {},
          ...(finalResponse == null ? void 0 : finalResponse.usage) ? {
            usage: {
              duration: finalResponse.usage.duration,
              outputVideoDuration: finalResponse.usage.output_video_duration,
              resolution: finalResponse.usage.SR,
              size: finalResponse.usage.size
            }
          } : {}
        }
      }
    };
  }
};

// src/version.ts
var VERSION = true ? "1.0.25" : "0.0.0-test";

// src/alibaba-provider.ts
function createAlibaba(options = {}) {
  var _a, _b;
  const baseURL = (_a = (0, import_provider_utils5.withoutTrailingSlash)(options.baseURL)) != null ? _a : "https://dashscope-intl.aliyuncs.com/compatible-mode/v1";
  const videoBaseURL = (_b = (0, import_provider_utils5.withoutTrailingSlash)(options.videoBaseURL)) != null ? _b : "https://dashscope-intl.aliyuncs.com";
  const getHeaders = () => (0, import_provider_utils5.withUserAgentSuffix)(
    {
      Authorization: `Bearer ${(0, import_provider_utils5.loadApiKey)({
        apiKey: options.apiKey,
        environmentVariableName: "ALIBABA_API_KEY",
        description: "Alibaba Cloud (DashScope)"
      })}`,
      ...options.headers
    },
    `ai-sdk/alibaba/${VERSION}`
  );
  const createLanguageModel = (modelId) => {
    var _a2;
    return new AlibabaLanguageModel(modelId, {
      provider: "alibaba.chat",
      baseURL,
      headers: getHeaders,
      fetch: options.fetch,
      includeUsage: (_a2 = options.includeUsage) != null ? _a2 : true
    });
  };
  const createVideoModel = (modelId) => new AlibabaVideoModel(modelId, {
    provider: "alibaba.video",
    baseURL: videoBaseURL,
    headers: getHeaders,
    fetch: options.fetch
  });
  const provider = function(modelId) {
    if (new.target) {
      throw new Error(
        "The Alibaba model function cannot be called with the new keyword."
      );
    }
    return createLanguageModel(modelId);
  };
  provider.specificationVersion = "v3";
  provider.languageModel = createLanguageModel;
  provider.chatModel = createLanguageModel;
  provider.video = createVideoModel;
  provider.videoModel = createVideoModel;
  provider.imageModel = (modelId) => {
    throw new import_provider4.NoSuchModelError({ modelId, modelType: "imageModel" });
  };
  provider.embeddingModel = (modelId) => {
    throw new import_provider4.NoSuchModelError({ modelId, modelType: "embeddingModel" });
  };
  return provider;
}
var alibaba = createAlibaba();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VERSION,
  alibaba,
  createAlibaba
});
